container-registry-configuration/README.md

```diff
@@ -0,0 +1,123 @@
+
+
+### ImagePullSecret
+
+We can set `imagePullSecrets` for all pods with an admission hook for authentication but Kubernetes doesn't have any mechanism like registry mirrors. That is why it is not a viable option.
+
+### FairwindsOps/saffire
+
+There is an open-source project to patch pods in `imagePullErrors` error state: https://github.com/FairwindsOps/saffire
+
+You are able to define alternative source for an image with a Custom Resource.
+
+```
+apiVersion: saffire.fairwinds.com/v1alpha1
+kind: AlternateImageSource
+metadata:
+  name: alternateimagesource-sample
+spec:
+  imageSourceReplacements:
+    - equivalentRepositories:
+        - quay.io/gianswarm/our-image
+        - azurecr.io/giantswarm/our-image
+        - docker.io/giantswarm/our-image
+```
+
+This mechanism doesn't work for static pods and it doesn't prevent the issue in advance. Also, it has no built-in mechanism for authentication. We can extend it or combine with the image pull secret solution but it sounds too complicated and dirty.
+
+
+### Containerd Configuration
+
+As we do in vintage clusters, we can configure containerd configuration as below. This seems the best approach.
+
+```
+[plugins."io.containerd.grpc.v1.cri".registry]
+
+[plugins."io.containerd.grpc.v1.cri".registry.mirrors]
+[plugins."io.containerd.grpc.v1.cri".registry.mirrors."docker.io"]
+endpoint = ["giantswarm.azurecr.io"]
+
+[plugins."io.containerd.grpc.v1.cri".registry.configs]
+[plugins."io.containerd.grpc.v1.cri".registry.configs."registry-1.docker.io".auth]
+username = "giantswarm-user-account"
+password = "my_token_from_docker_hub"
+```
+
+## Registry usage
+
+### Which registry should we use? 
+
+The infranet page (See `Registry Mirrors`) states this as a desired configuration:
+  - `docker.io` as primary public registry since it has a privilege in docker daemon.
+  - Our own Azure Container Registry as secondary registry because of the security concerns
+
+Using only public images without authentication is also on the table. In that case, we will be limited by total of [free account limit](https://docs.docker.com/docker-hub/download-rate-limit/) of Docker Hub (100 pulls per 6 hours per IP address) and [tier limit](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-skus) of Azure Container Registry (3000 ReadOps per minute for standard). However, anyone can pull images from our registry and spend our limits, which makes us vulnerable. That is why we must use authenticated users.  
+
+### Open Points
+
+## How are we going to configure containerd in cluster app?
+
+We need to provide a configation interface in our `cluster-$provider` apps for the users so that they can provide their credentials. Note that we are the users of our management clusters.
+
+Options:
+1. We can ask users to create whole containerd configuration as a secret.
+2. We can ask users to provide list of registries and credentials as a secret and render containerd configuration with them.
```

Comment:
Why would we do this rather than making use of `imagePullSecrets` where the customer then has complete control over what uses their creds?

Reply:
We need credentials for control-plane elements too. We need to put them into container configuration for static pods.  Should we use separate 2 credentials for control-plane&user workloads? 

Note: I am going to update this RFC soon according to the results of yesterday's kaas sync.